sgkit_vcf/vcf_reader.py

```diff
@@ -0,0 +1,286 @@
+                # Also chunk in the samples direction
+                encoding = dict(
+                    call_genotype=dict(chunks=(chunk_length, chunk_width, n_ploidy)),
+                    call_genotype_mask=dict(
+                        chunks=(chunk_length, chunk_width, n_ploidy)
+                    ),
+                    call_genotype_phased=dict(chunks=(chunk_length, chunk_width)),
+                    variant_allele=dict(chunks=(chunk_length, n_allele)),
+                    variant_contig=dict(chunks=(chunk_length,)),
+                    variant_id=dict(chunks=(chunk_length,)),
+                    variant_id_mask=dict(chunks=(chunk_length,)),
+                    variant_position=dict(chunks=(chunk_length,)),
+                    sample_id=dict(chunks=(chunk_width,)),
+                )
+
+                ds.to_zarr(output, mode="w", encoding=encoding)
+                first_variants_chunk = False
+            else:
+                # Append along the variants dimension
+                ds.to_zarr(output, append_dim=DIM_VARIANT)
+
+
+def vcf_to_dataset(
+    input: PathType,
+    output: PathType,
+    region: Optional[str] = None,
+    chunk_length: int = 10_000,
+    chunk_width: int = 1_000,
+) -> xr.Dataset:
+    vcf_to_zarr_sequential(input, output, region, chunk_length, chunk_width)
+    ds: xr.Dataset = xr.open_zarr(str(output))  # type: ignore[no-untyped-call]
+    return ds
+
+
+def vcf_to_zarr_parallel(
+    input: Union[PathType, Sequence[PathType]],
+    output: PathType,
+    regions: Union[None, Sequence[str], Sequence[Optional[Sequence[str]]]],
+    chunk_length: int = 10_000,
+    chunk_width: int = 1_000,
+) -> None:
+    """Convert specified regions of one or more VCF files to zarr files, then concat, rechunk, write to zarr"""
+
+    if isinstance(input, str) or isinstance(input, Path):
+        # Single input
+        inputs: Sequence[PathType] = [input]
+        assert regions is not None  # this would just be sequential case
+        input_regions: Sequence[Optional[Sequence[str]]] = [regions]  # type: ignore
+    else:
+        # Multiple inputs
+        inputs = input
+        if regions is None:
+            input_regions = [None] * len(inputs)
+        else:
+            if len(regions) == 0 or isinstance(regions[0], str):
+                raise ValueError(
+                    f"Multiple input regions must be a sequence of sequence of strings: {regions}"
+                )
+            input_regions = regions
+
+    assert len(inputs) == len(input_regions)
+
+    tempdir = Path(tempfile.mkdtemp(prefix="vcf_to_zarr_"))
+
+    datasets = []
+    parts = []
+    for i, input in enumerate(inputs):
+        filename = Path(input).name
+        input_region_list = input_regions[i] or [None]  # type: ignore
+        for r, region in enumerate(input_region_list):
+            part = tempdir / filename / f"part-{r}.zarr"
+            parts.append(part)
+            ds = dask.delayed(vcf_to_dataset)(
```

Comment:
I don't really understand the implications of it, but I know dask generally recommends not making [delayed tasks within delayed tasks](https://docs.dask.org/en/latest/delayed-best-practices.html#avoid-calling-delayed-within-delayed-functions) where the `ds.to_zarr` in `vcf_to_zarr_sequential` is the inner task here.  I don't see an easy way to avoid that but I thought I'd mention it in case you do.

Reply:
Unless I'm missing something, `ds.to_zarr` in `vcf_to_zarr_sequential` runs immediately so I don't think this is a problem. It only returns a delayed object if `compute=False`, which it never is here.

Reply:
Hm I guess there are two cases:

1. A delayed task runs one more complete task graphs in isolation
2. A delayed task returns one or more delayed tasks

I had thought both were bad with a vague sense that 2 is worse than 1, yet neither seems to really matter when using a local scheduler.  Sounds good to me to ignore that until there's a clearer reason that it matters.